engine/lib/pagehandler.php

```diff
@@ -58,6 +58,101 @@ function elgg_generate_url($name, array $parameters = []) {
 	return _elgg_services()->router->generateUrl($name, $parameters);
 }
 
+/**
+ * Generate entity URL from a named route
+ * Route segments will be automatically resolved from entity attributes and metadata
+ *
+ * @param ElggEntity $entity      Entity
+ * @param string     $resource    Resource name
+ * @param string     $subresource Subresource name
+ * @param array      $query       URL query elements
+ *
+ * @return string
+ */
+function elgg_generate_entity_url(ElggEntity $entity, $resource = 'view', $subresource = null, $query = []) {
+
+	$make_route_name = function($type, $subtype) use ($resource, $subresource) {
+		$route_parts = [
+			$resource,
+			$type,
+			$subtype,
+			$subresource,
+		];
+
+		return implode(':', array_filter($route_parts));
+	};
+
+	$pairs = [
+		[$entity->getType(), $entity->getSubtype()],
+		[$entity->getType(), null],
+	];
+
+	$route = false;
+	$route_name = '';
+
+	foreach ($pairs as $pair) {
+		$route_name = $make_route_name($pair[0], $pair[1]);
+		$route = _elgg_services()->routeCollection->get($route_name);
+		if ($route) {
+			break;
+		}
+	}
+
+	if (!$route) {
+		return;
+	}
+
+	$requirements = $route->getRequirements();
+	$defaults = $route->getDefaults();
+
+	$props = array_merge(array_keys($requirements), array_keys($defaults));
+
+	foreach ($props as $prop) {
+		if (substr($prop, 0, 1) === '_') {
+			continue;
+		}
+
+		if (isset($query[$prop])) {
+			continue;
+		}
+
+		switch ($prop) {
+			case 'title' :
+			case 'name' :
+				$query[$prop] = elgg_get_friendly_title($entity->getDisplayName());
+				break;
+
+			default :
+				$query[$prop] = $entity->$prop;
+				break;
+		}
+	}
+
+	return elgg_generate_url($route_name, $query);
+}
+
+/**
+ * Generate an action URL
+ *
+ * @param string $action          Action name
+ * @param array  $query           Query elements
+ * @param bool   $add_csrf_tokens Add tokens
+ *
+ * @return string
+ */
+function elgg_generate_action_url($action, array $query = [], $add_csrf_tokens = true) {
+
+	$url = "action/$action";
```

Comment:
Yes! Never wondering if it's "actions/" or "action/".

Reply:
Not to mention `"action/friends/add?guid=" . elgg_get_page_owner_guid()`